pennylane_lightning/lightning_tensor/_tensornet.py

```diff
@@ -223,20 +221,35 @@ def _preprocess_state_vector(self, state, device_wires):
         if len(device_wires) == self._num_wires and Wires(sorted(device_wires)) == device_wires:
             return np.reshape(state, output_shape).ravel(order="C")
 
-        # generate basis states on subset of qubits via the cartesian product
-        basis_states = np.array(list(product([0, 1], repeat=len(device_wires))))
+        local_dev_wires = np.array(device_wires.tolist())[::-1]
+
+        # generate basis states on subset of qubits via broadcasting
+        base = np.tile([0, 1], 2 ** (len(local_dev_wires) - 1)).astype(dtype=np.int64)
+        indexes = np.zeros(2 ** (len(local_dev_wires)), dtype=np.int64)
+
+        max_dev_wire = self._num_wires - 1
 
         # get basis states to alter on full set of qubits
-        unravelled_indices = np.zeros((2 ** len(device_wires), self._num_wires), dtype=int)
-        unravelled_indices[:, device_wires] = basis_states
+        for i, wire in enumerate(local_dev_wires):
+
+            # get indices for which the state is changed to input state vector elements
+            indexes += base * 2 ** (max_dev_wire - wire)
+
+            if i == len(local_dev_wires) - 1:
+                continue
+
+            two_n = 2 ** (i + 1)
+            base = base.reshape(-1, two_n * 2)
+            swaper_A = two_n // 2
+            swaper_B = swaper_A + two_n
```

Comment:
I guess it's not fully clear to me why we are doing what we are doing here? Could you walk me through it?

Reply:
An explanation of the change:

The current algorithm is the following:

1. Generate basis states on a subset of qubits via the cartesian product
``` python
basis_states = np.array(list(product([0, 1], repeat=len(dev_wires))))
```
The line above produces the following pattern (Transposed)
```
[[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
 [0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1]
 [0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1]
 [0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1]
 [0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1]]
```
 For a small system, it is not a problem. However, here we are creating a list of `2 ** dev_wires * dev_wires` which will overload the memory if we try to work with 30 qubits. 

2. Create a new array with size `2 ** dev_wires * num_wires` to allocate each basis row to the `state prep` row, producing a second big allocation.  

``` python
unravelled_indices = np.zeros((2 ** len(device_wires), self._num_wires), dtype=int)
unravelled_indices[:, device_wires] = basis_states
unravelled_indices.T
```

3.  Multiply each row by `2**n` where `n` goes from `num_wires` to 1 and then sum each column to get indices for which the state is changed to input state vector elements. This can be done with the following line:

``` python
ravelled_indices = np.ravel_multi_index(unravelled_indices.T, [2] * self._num_wires)
```

4. With the indices from `ravelled_indices` the new state `full_state` is filled. 
``` python
 full_state = np.zeros(2**self._num_wires, dtype=self.dtype)
 for i, value in enumerate(state):
    full_state[ravelled_indices[i]] = value
```

The solution that we propose is the following: 
 
1. Create a single row of  `len(device_wires)` length with the pattern `[0 1 0 1 ....]`.  
2. Multiply by the basis `2**n`  and accumulate the value of the indices in the array `indexes`. 
3. Start a for loop for the next row of `base`. Instead of reallocating the `base` array, it is faster to swap columns because it only changes the viewer for the array.  
4. With the indices from `indexes` the new state `full_state` is filled. 
``` python 
        base = np.tile([0, 1], 2 ** (len(local_dev_wires) - 1)).astype(dtype=np.int64)
        indexes = np.zeros(2 ** (len(local_dev_wires)), dtype=np.int64)

        for i, wire in enumerate(local_dev_wires):
            two_n = 2 ** (i + 1)
            indexes += base * 2 ** (max_dev_wire - wire)
            
            if i == len(local_dev_wires) - 1:
                continue

            base = base.reshape(-1, two_n * 2)
            swaper_A = two_n // 2
            swaper_B = swaper_A + two_n

            base[:, swaper_A:swaper_B] = base[:, swaper_A:swaper_B][:, ::-1]
            base = base.reshape(-1)
```

The swap works as follows:
1. From a flattened array, it reshaped to a number of columns of `2**(n+1)` 
2. Swapped the internal range from `2**(n)` to `2**(n+2)` 
3. Flatten the matrix. 

This is an example of the evolution of the algorithm 
```
First allocation:
[0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1]

Reshape
[[0 1 0 1]
 [0 1 0 1]
 [0 1 0 1]
 [0 1 0 1]
 [0 1 0 1]
 [0 1 0 1]
 [0 1 0 1]
 [0 1 0 1]]

Swap internal columns
[[0 0 1 1]
 [0 0 1 1]
 [0 0 1 1]
 [0 0 1 1]
 [0 0 1 1]
 [0 0 1 1]
 [0 0 1 1]
 [0 0 1 1]]

Flatten
[0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1 0 0 1 1]

Start again. Reshape
[[0 0 1 1 0 0 1 1]
 [0 0 1 1 0 0 1 1]
 [0 0 1 1 0 0 1 1]
 [0 0 1 1 0 0 1 1]]

Swap internal columns
[[0 0 0 0 1 1 1 1]
 [0 0 0 0 1 1 1 1]
 [0 0 0 0 1 1 1 1]
 [0 0 0 0 1 1 1 1]]

Flatten
[0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1 0 0 0 0 1 1 1 1]
....
```

Reply:
I will add more documentation to the code to make more readable this change.

Reply:
Thanks for the explanation, this is really clear.

As a follow on, do we even need the single allocation, single we are seemingly just generating parity/gray patterns iteratively? Couldn't we just build these on the fly for each iteration and map/select the values without building the full array?

Reply:
Yes could be a good approach but I think this will be slower because we will use one or two extra Python `for` loops to compute the value for the `indexes` but I am not sure. I suggest keeping this implementation and if in the future we need to optimize this section then apply this iteration on the fly. What do you think? :thinking:

Reply:
Sounds good. I think it'll be faster at a certain threshold, as you do less memory ops than the columnar swaps, but that could be anywhere in the 21-27 qubits range. Are you ok to track a backlog story to investigate this approach at a later date?